src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp

```diff
@@ -53,18 +53,99 @@ namespace OpenMS
   {
     const double intensity_l = chromatogram.RTBegin(left)->getIntensity();
     const double intensity_r = (chromatogram.RTEnd(right)-1)->getIntensity();
-    const double delta_int = intensity_r - intensity_l; // sign will determine line direction
-    const double delta_input_rt = right - left;
+    const double delta_int = intensity_r - intensity_l;
     double background = 0.0;
-    for (auto it=chromatogram.RTBegin(left); it!=chromatogram.RTEnd(right); ++it)
+    if (integration_type_ == "trapezoid" || integration_type_ == "simpson")
     {
-      // calculate the background using the formula
-      // y = mx + b where x = retention time, m = slope, b = left intensity
-      background += delta_int / delta_input_rt * (it->getRT() - left) + intensity_l;
+      const double intensity_min = std::min(intensity_r, intensity_l);
+      const double delta_int_abs = std::fabs(delta_int);
+      for (auto it=chromatogram.RTBegin(left); it<chromatogram.RTEnd(right)-1; ++it)
+      {
+        const double delta_rt = (it+1)->getRT() - it->getRT();
+        background += intensity_min * delta_rt + 0.5 * delta_int_abs * delta_rt;
+      }
+    }
+    else
+    {
+      const double delta_input_rt = right - left;
+      for (auto it=chromatogram.RTBegin(left); it!=chromatogram.RTEnd(right); ++it)
+      {
+        // calculate the background using the formula
+        // y = mx + b where x = retention time, m = slope, b = left intensity
+        // sign of delta_int will determine line direction
+        background += delta_int / delta_input_rt * (it->getRT() - left) + intensity_l;
+      }
     }
     return background;
   }
 
+  void PeakIntegrator::integratePeak(
+    const MSChromatogram& chromatogram,
+    const double& left,
+    const double& right,
+    double& peak_area,
+    double& peak_height,
+    double& peak_apex_pos
+  )
+  {
+    peak_area = 0.0;
+    peak_height = -1.0;
+    peak_apex_pos = -1.0;
+    UInt n_points = 0;
+    for (auto it=chromatogram.RTBegin(left); it!=chromatogram.RTEnd(right); ++it, ++n_points)
+      ;
+
+    if (getIntegrationType() == "trapezoid")
+    {
+      for (auto it=chromatogram.RTBegin(left); it!=chromatogram.RTEnd(right)-1; ++it)
+      {
+        peak_area += ((it+1)->getRT() - it->getRT()) * ((it->getIntensity() + (it+1)->getIntensity()) / 2.0);
+        if (peak_height < it->getIntensity())
+        {
+          peak_height = it->getIntensity();
+          peak_apex_pos = it->getRT();
+        }
+      }
+    }
+    else if (getIntegrationType() == "simpson")
+    {
+      if (n_points < 3 || !(n_points % 2))
```

Comment:
I can see this part of the if clause causing a lot of headache: `!(n_points % 2)`.  Would it be possible for the case that n_points is even to calculate the peak area with 1) 1 additional point on the right and 2) with 1 less point on the right and then 3) take the average to approximate what the area is?

Reply:
Sure, we can do this, assuming there is a `Peak` after `right` position (I guess it's safe to assume so)